resources/testdriver-actions.js

```diff
@@ -0,0 +1,383 @@
+(function() {
+  let sourceNameIdx = 0;
+
+  /**
+   * Builder for creating a sequence of actions
+   */
+  function Actions() {
+    this.sourceTypes = new Map([["key", KeySource],
+                                ["pointer", PointerSource],
+                                ["general", GeneralSource]]);
+    this.sources = new Map();
+    for (let sourceType of this.sourceTypes.keys()) {
+      this.sources.set(sourceType, new Map());
+    }
+    this.currentSources = new Map();
+    for (let sourceType of this.sourceTypes.keys()) {
+      this.currentSources.set(sourceType, null);
+    }
+    this.createSource("general");
+    this.tickIdx = 0;
+  }
+
+  Actions.prototype = {
+    /**
+     * Generate the action sequence suitable for passing to
+     * test_driver.action_sequence
+     *
+     * @returns {Array} Array of WebDriver-compatible actions sequences
+     */
+    serialize: function() {
+      let actions = [];
+      for (let [_, sources] of this.sources) {
+        for (let [sourceName, source] of sources) {
+          let serialized = source.serialize(this.tickIdx + 1);
+          if (serialized) {
+            serialized.id = sourceName;
```

Comment:
Just to double check here since the current implementation does seem so support this scenario and as long as the driver supports this is should work. The developer can call pointerdown for say a touch source with name "finger 1" in one call to Actions and send it. Then they should be able to create the same source (with the same name and pointertype) in the next call and this time add a pointerup. Correct?

Reply:
I just implemented this.

Reply:
I believe what you already have is perfect. Particularly I believe most of the use cases don't care about multiple sources at the same time. So what we have is fine and concise and is capable of representing everything already. As you suggested we can always enhance this later if we really think it is needed.